Spots final

coop0183 opened the pull request.

coop0183 committed:
updated package.json

coop0183 committed:
modified:   src/pages/index.js
	new file:   src/utils/Api.js

coop0183 committed:
modified:   src/blocks/modal.css
	modified:   src/blocks/profile.css
	new file:   src/images/pencil-light.svg
	modified:   src/index.html
	modified:   src/pages/index.js
	modified:   src/utils/Api.js

coop0183 committed:
modified:   src/blocks/modal.css
	modified:   src/index.html
	modified:   src/pages/index.js
	modified:   src/utils/Api.js

coop0183 committed:
modified:   src/pages/index.js
	modified:   src/utils/Api.js

coop0183 committed:
modified:   src/pages/index.js

coop0183 committed:
modified:   src/pages/index.js

coop0183 committed:
modified:   src/pages/index.js

coop0183 committed:
modified:   src/pages/index.js
	modified:   src/utils/Api.js

gennady-bars commented on src/index.html:
```diff
@@ -189,5 +198,65 @@ <h2 class="modal__title">New Post</h2>
+                        Profile Picture
+                        <input
+                            id="profile-avatar-input"
+                            type="text"
```
* Please, add `type="url"` for the `avatar` input, to validate it as a `link`

gennady-bars commented on src/utils/Api.js:
```diff
@@ -0,0 +1,87 @@
+            return res.json();
+        }
+        return Promise.reject(`Error: ${res.status}`);
+    }
```
**COULD BE IMPROVED**

You can make a universal `_request`  method  to avoid duplication of  the `ok` checking, `headers` and `baseUrl`


```js
 _request(endpoint, options = {}) {
    const finalOptions = {
      // add the default headers
      headers: this._headers,
      // spread the other options if provided
      ...options,
    };
      // add the endpoint to the base url
    const url = `${this._baseUrl}${endpoint}`;
    return fetch(url, finalOptions).then(this._checkResponse);
  }

  // usage

  getUserInfo() {
    // pass only the endpoint here
    return this._request("/users/me");
  }

  getInitialCards() {
    // pass only the endpoint here
    return this._request("/cards");
  }

  // and so on



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -108,6 +167,40 @@ function getCardElement(data) {
+            if (data) {
+                evt.target.classList.toggle("card__like-btn_active");
+            }
+            localStorage.setItem("likedCard", cardId);
```
`            localStorage.setItem("likedCard", cardId);
`

`localStorage` is not needed in your project because you should get any info from the **server**. Please, delete it

gennady-bars commented on src/pages/index.js:
```diff
@@ -153,37 +250,77 @@ document.querySelectorAll(".modal__close-btn").forEach((button) => {
 editProfileModal.addEventListener("submit", function (evt) {
```
**COULD BE IMPROVED**

If it’s interesting for you, here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **resetting** and **catching errors**


```js
// define a function for changing the button text. It accepts 4 params (the 2 last are optional with default texts)
export function renderLoading(isLoading, button, buttonText='Save', loadingText='Saving...') {
  if (isLoading) {
    button.textContent = loadingText
  } else {
    button.textContent = buttonText
  }
}

// define a universal function that accepts a request function, event and a default loading text 
function handleSubmit(request, evt, loadingText = 'Saving...') {
 // You need to prevent the default action in any submit handler
  evt.preventDefault();

  // the button is always available inside `event` as `submitter`
  const submitButton = evt.submitter;
  // fix the initial button text
  const initialText = submitButton.textContent;
  // change the button text before requesting
  renderLoading(true, submitButton, initialText, loadingText);
  // call the request function to be able to use the promise chain
  request()
    .then(() => {
      // any form should be reset after a successful response 
      // evt.target is the form in any submit handler		
      evt.target.reset();
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
    .catch(console.error)      

    // and in finally we need to stop loading
    .finally(() => {
      renderLoading(false, submitButton, initialText);
    });
}

```

Here is an example of handling the profile form submit:

```js
function handleProfileFormSubmit(evt) {
  // create a request function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return editProfile(nameInput.value, jobInput.value).then((userData) => {
      userName.textContent = userData.name;
      userJob.textContent = userData.about;
    });
  }
  // here we call handleSubmit passing the request and event (if you want a different loading text then you need to pass the 3rd param)
  handleSubmit(makeRequest, evt);
}

```

So, this way you can remove a lot of code duplicating. You will not need to search buttons, pass initial button texts and so on.

`handleSubmit` and `renderLoading` should be placed in `utils.js`, because they are utility functions

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Jonathon!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

Please, scroll down the page on **github** to see my remarks right in the code

**Parts I really liked in your project:**

* Good job that you **close** modals only in `.then` blocks
* Good job that you **catch** possible errors at the end of server requests.
* Good job that you return the button text in `.finally`

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:
* Please, disable the `avatar` save button after submission as well
* Please, add `type="url"` for the `avatar` input, to validate it as a `link`


There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

coop0183 committed:
modified:   src/index.html
	modified:   src/pages/index.js
	modified:   src/utils/Api.js
	new file:   src/utils/utils.js

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -153,37 +244,78 @@ document.querySelectorAll(".modal__close-btn").forEach((button) => {
+            }
+            closeModal(editProfileModal);
+            editProfileForm.reset();
+            disabledBtn(editProfileBtn, validationConfig);
```
Currently, the button that opens the Edit Profile modal is being disabled after a successful submission, which prevents the modal from being opened again.

To fix this, make sure you're disabling the submit button inside the form after a successful submission. You can do that by updating this line to:

```js
disableButton(submitBtn, validationConfig);
```

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -153,37 +244,78 @@ document.querySelectorAll(".modal__close-btn").forEach((button) => {
+                profileDescriptionEl.textContent = data.about;
+            }
+            closeModal(editProfileModal);
+            editProfileForm.reset();
```
**COULD BE IMPROVED**

It's not necessary to reset the Edit Profile form after submission because the input fields are automatically filled with the user's current profile data when the modal opens. Therefore, this line can be safely removed to simplify the code.

sebasLTripleTen reviewed (changes requested):
Hi!

Thank you for considering our recommendations. Your project has definitely improved. Most of the issues have been resolved, but there is something that still need attention:

- Currently, the Edit Profile button is being disabled after a successful submission, which prevents the modal from being opened again.

Once this minor issue is addressed, your project will be ready for approval!

coop0183 committed:
modified:   src/pages/index.js

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
